### DISCUSSION TITLE
Concerns regarding UTF8 string literals proposal

### DISCUSSION THREAD

**Discussion by DaZombieKiller** [dotnet/csharplang#5983] [Language Ideas]
There was a discussion in the C# Discord regarding the UTF8 string literals proposal, and a number of concerns were raised:

### `byte[]` will result in unintentional allocations as the natural type.

This would result in `var buffer = "hello"u8;` being a "hidden" allocation, which is arguably unintuitive as it differs from the `string` equivalent and is _very_ easy to skim over in review. `ReadOnlySpan<byte>` would be a more suitable natural type for the literal.

### UTF8 literals should be null-terminated.

UTF8 literals are highly likely to see heavy use in low-level and interop-related code, where null-terminated strings are often expected.

The [LDM notes](https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-10-27.md#utf-8-string-literals) from October 27th, 2021 state the following reasons for UTF8 literals _not_ being null-terminated currently:

> * Should the byte sequence have a null terminator?
>     * Interop scenarios will want a null terminator. However, they can add one by including a `\0` at the end of the string, and an analyzer can catch these cases.
>     * On the other hand, most managed code scenarios would break if a terminator was included.

Manually adding a null terminator to the literal is very easy to forget, and I'm not personally confident that an analyzer would be able to catch all cases where one is necessary. It is not possible for it to accurately determine when a terminator is necessary every time, and the (rather unappetizing) alternative is to flag _every_ callsite.

The point about managed code scenarios breaking would no longer be a concern if the natural type of the literal were to become `ReadOnlySpan<byte>`, with the terminator sliced out of the span's length.

### Implicit conversions to `Span<byte>` and `byte[]` will allocate.

Similarly to the initial concern regarding the natural type, these are hidden allocations.

Today, if you wish to get a `char[]` buffer from a `string` literal, you can use the `ToCharArray` method:
```cs
char[] buffer = "hello".ToCharArray();
```
If the natural type of a UTF8 literal is changed to `ReadOnlySpan<byte>`, the result is a very similar API which does not hide allocations:
```cs
byte[] buffer = "hello"u8.ToArray();
Span<byte> span = "hello"u8.ToArray();
```

### Inconsistencies with `string`.

The proposal allows for the following:
```cs
var array1 = "hello"u8; // byte[]
byte[] array2 = "hello";
ReadOnlySpan<byte> span1 = "hello";
Span<byte> span2 = "hello";
```
But does not propose equivalents for strings:
```cs
char[] array = "hello";
Span<char> span = "hello";
```
This creates an inconsistency between the two types of literals.

**Comment by tannergooding** [comment]
CC. @MadsTorgersen, @stephentoub 

Feedback/Questions about the UTF8 String Literal Design (https://github.com/dotnet/csharplang/issues/184)

**Comment by stephentoub** [comment]
> UTF8 literals are highly likely to see heavy use in low-level and interop-related code, where null-terminated strings are often expected.

What does the interop code you're planning to write with these look like?

I don't have a huge problem if we wanted to emit an extra null byte at the end of the data section (not included in the returned span itself), but I do wonder how useful that'd actually be in wide-spread practice, and I am concerned that it might actually lead to a larger pit of failure.  With strings, a developer can trust that any `System.String` they're handed has a null terminator after the last char of the string, so interop code can be written to work with arbitrary string instances passed in.  That is _not_ the case for `ReadOnlySpan<char>` or `ReadOnlySpan<byte>`, which in general are not guaranteed to be followed by a null terminator, and very often won't be.  As such, you can't write a similar interop routine that accepts an arbitrary `ReadOnlySpan<byte>` and expect it to be followed by a null terminator, nor should you look beyond the end of a `ReadOnlySpan<byte>`'s length to see if it's followed by a null.  So, while there might be benefits for _some_ use cases (e.g. where your interop code directly references the UTF8 span rather than it working with arbitrary spans), I expect it won't be anywhere close to as common as with strings, and that the language trying to make guarantees here could actually backfire, if less experienced developers come to think that `ReadOnlySpan<byte>`s are null terminated, write code expecting it, and then things start failing spectacularly when passed a slice or a span from some array or most any other data source.  I've seen that exact issue multiple times with spans created from strings, where the span tested with happened to be null terminated because it was just a `"test string"` passed in, and then things fell over when slices or cast arrays or stackalloc'd data or so on were passed in.

**Reply by DaZombieKiller** [reply]
> What does the interop code you're planning to write with these look like?

For the most part, instances like this:
```cs
fixed (byte* pText = "LiteralString"u8)
{
    SomeNativeMethod(pText);
}
```
But also scenarios where I may want to pass the string around or accept it as user input. That currently could only be passed as `ReadOnlySpan<byte>`, which is indeed a bit of a problem:

> That is not the case for `ReadOnlySpan<char>` or `ReadOnlySpan<byte>`, which in general are not guaranteed to be followed by a null terminator, and very often won't be.

If anything I think it makes a stronger argument for a dedicated `Utf8String` type. `ReadOnlySpan<byte>` is only really involved in this particular concern due to the lack of such a type.

**Reply by stephentoub** [reply]
> If anything I think it makes a stronger argument for a dedicated Utf8String type

Even if a `Utf8String` is eventually introduced (and that is a big "if"), it will definitely not be in .NET 7.

**Reply by tannergooding** [reply]
> but I do wonder how useful that'd actually be in wide-spread practice, and I am concerned that it might actually lead to a larger pit of failure. 

I don't think it will and actually think the current setup is more of a pit of failure.

In either setup, whether `utf-8` and `utf-16` string literals work the same (both null-terminated) or differently (only utf-16 string literals are null terminated) then the user given an arbitrary `T[]`, `Span<T>`, or `ReadOnlySpan<T>` (where `T` is `byte` or `char`) cannot assume it is null terminated and will have to allocate data that is +1 bigger (to hold the null-terminating character), copy the data over, and set the final character to `\0`.

On the above, users already have a "potential" pit of failure in that if they assume it is null-terminated it will work for `utf-16 string literals` and not work for substrings. I don't think `utf-8 string literals` having the same "potential" pit of failure is worth making these differ. I think its consistent and means users already doing interop will know to catch and handle it (familiar users already know this step "squeaks"). Users using the `DllImport Source Generator` will likewise have the marshalling implicitly doing the right thing here. Consistency is good and users will understand or be able to infer the "same pit" exists from existing SO posts and other discussions on the topic that have been around for 10-20 years.

What remains is how `"string"` and `"string"u8` when consumed directly work. If they work consistently then both are null-terminated and users have nothing "new" to learn. However, if they differ then rather than users needing to deal with the "same well known pit", they have to watch out for several new pits of failure (now rather than 1 consistent step that "squeaks", several additional steps may "squeak"). Namely:
* For `utf-8 string literal` they have to postfix with `\0`
* If they also need the length, they then need to remember to subtract one to get the relevant length
* If they forget the `\0` then the "worst" case is "unbounded" (at least until the page changes at which point it may or may not AV) memory buffer access which is going to cause a CVE
  * Where-as if we null-terminate, then the "worst" case is the same as for `string`. It may read past the end of the substring but only up to the end of the backing string literal. This is still not great, but having it deal with `"Hello, World"` rather than just the intended `"Hello"` is much less problematic than dealing with `"Hello, World<random bytes of data, who knows what it is, maybe we'll hit a null>"`

**Reply by Joe4evr** [reply]
> much less problematic than dealing with `"Hello, World<random bytes of data, who knows what it is, maybe we'll hit a null>"`

![image](https://user-images.githubusercontent.com/1058562/161678913-70f20825-f5ed-4a8c-8d35-7b8834333519.png)

**Reply by LokiMidgard** [reply]
> is much less problematic

Not sure about that. I find it much easyer to see what the problem is having the later value.

**Comment by roji** [comment]
The [alternatives section of the current proposal](https://github.com/dotnet/csharplang/blob/main/proposals/utf8-string-literals.md#alternatives) doesn't mention one possibility: having the compiler identify `Encoding.UTF8.GetBytes(literal)` as a special case. In other words, simply make the following two equivalent:

```c#
static ReadOnlySpan<byte> BytesStatus100 => Encoding.UTF8.GetBytes("100");
static ReadOnlySpan<byte> BytesStatus100 => new byte[] { (byte)'1', (byte)'0', (byte)'0' };
```

This seems like it would deliver the advantages of the proposed new UTF8 literal, but without any language modifications, and without having to deal with any of the complexities of the proposal (e.g. resolution overload, what happens in expression trees...). It is also more explicit than the current proposal, not leaving any room for doubt as to which byte representation is returned:

```c#
// Since "hello" is a (UTF16) string, it's reasonable for someone to assume that array contains the UTF16 byte representation, not the UTF8 one.
byte[] array = "hello";
```

I'm guessing this was considered and rejected for good reasons, but it would be good to document those reason in the alternatives section of the proposal.

**Reply by tannergooding** [reply]
This may be missing that `static ReadOnlySpan<byte> X => new byte[] { cns, cns, cns };` is specialized by the compiler and actually emits a PE embedded constant avoiding all allocations here.

This would mean that the current implementation is "free" while `Encoding.UTF8.GetBytes` would not participate in constant folding and would require allocations among other work for it to behave as expected.

**Reply by teo-tsirpanis** [reply]
By "compiler" I presume roji meant _Roslyn_, not the JIT.

**Reply by tannergooding** [reply]
Yes. Roslyn optimizes `new byte[] { cns, cns, cns }` (and same for `sbyte`) to be a `.data` declaration in the IL, which is effectively an embedded PE constant.

When that is directly turned into a `ROSpan<byte>`, the compiler is able to optimize away the allocation all together and just have the `ROSpan<byte>` directly track the PE constant.

When its turned into a `Span<byte>` or `byte[]`, the compiler will allocate but then efficiently initialize the array via a direct copy of the backing PE constant bytes.

There is a similar optimization available for the other primitive types, but they always allocate and copy the backing PE constant bytes since they may have to adjust for endianness. There is a separate runtime proposal for an API that would allow the language to avoid the allocation for `ROSpan<T>` on the other primitive types.

**Reply by teo-tsirpanis** [reply]
We are in agreement on that @tannergooding. What I don't understand is the following:

> `Encoding.UTF8.GetBytes` would not participate in constant folding and would require allocations among other work for it to behave as expected.

`GetBytes` won't be called at all in IL. It would be a special-cased pattern for when the string parameter is a constant and the returned array is used as a read-only span, exactly like the pattern you described when we get a span from a constant `new byte []` but `newarr` is not emitted at all.

Which is not a bad idea, but Roslyn does not do cross-assembly optimizations.

**Reply by tannergooding** [reply]
Right, it would be a form of `constexpr` support, which the language doesn't broadly support.

**Reply by roji** [reply]
@tannergooding I'm sure I'm missing something, but I don't quite understand why Roslyn can't do exactly the same thing when it sees this:

```c#
static ReadOnlySpan<byte> BytesStatus100 => Encoding.UTF8.GetBytes("100");
```

... as when it sees this:

```c#
static ReadOnlySpan<byte> BytesStatus100 => new byte[] { (byte)'1', (byte)'0', (byte)'0' };
```

In other words, Roslyn already knows to identify a special code pattern (new byte[] in this context), and optimize it to be a .data declaration in the IL. Why could it not do the exact same thing for the expression `Encoding.UTF8.GetBytes()` in the same context, also emitting a .data declaration?

More broadly, the goals of this proposal *seem* like they can be achieved without any language spec changes (and with better explicitness as well) - if so, that would surely be a better way of achieveing them? Is there some Roslyn implementation issue blocking this?

**Comment by teo-tsirpanis** [comment]
Another incinsistency with strings I ~~noticed~~ guessed is around overload resolution. Let's say we have the following code:

``` csharp
public static class Parser {
    public static void ParseUtf16(ReadOnlySpan<char> span) {}
    public static void ParseUtf16(ReadOnlyMemory<char> span) {}
    public static void ParseUtf8(ReadOnlySpan<byte> span) {}
    public static void ParseUtf8(ReadOnlyMemory<byte> span) {}
}
```

If I call `Parser.ParseUtf16("Hello")`, the compiler will prefer the `ReadOnlySpan<char>` overload. What would happen if I call `Parser.ParseUtf8("Hello"u8)`? My guess is that since the UTF-8 literals' natural type is a byte array, it would have the same effect as calling `Parser.ParseUtf8(new byte[] {...})`, which throws an error because the compiler cannot pick an overload. For consistency and performance, `Parser.ParseUtf8("Hello"u8)` should prefer the `ReadOnlySpan<byte>` overload, and when and if `Utf8String` becomes available and such overload gets introduced, it should prefer that.

**Reply by tannergooding** [reply]
Related: https://github.com/dotnet/roslyn/issues/60612

**Reply by teo-tsirpanis** [reply]
I actually tried it with SharpLab and `Parser.ParseUtf8("Hello"u8)` failed as I expected.

However, `Parser.ParseUtf8("Hello")` _without the `u8`_ compiles and calls the `ReadOnlySpan<byte>` overload.

**Comment by teo-tsirpanis** [comment]
According to the proposal, the literal's natural type was chosen to be `byte[]` because spans are not available on .NET Framework.

Why not make it a `ReadOnlySpan<byte>` and _fail_ if the compiler cannot find the type (requiring the `System.Memory` NuGet package)? C# versions after 7.3 are not officially supported on .NET Framework and .NET Standard 2.0, and we shouldn't constrain future language decisions because of them. And the feature would still work, just with a polyfill.

**Reply by teo-tsirpanis** [reply]
One complication with my proposal is that we could not use UTF-8 literals in `byte[]`-typed attribute arguments, unless the compiler special-cases `[MyAttribute("Hello"u8.ToArray())]`.
